src/libraries/Common/src/System/Text/Json/PooledByteBufferWriter.cs

```diff
@@ -1,249 +1,97 @@
-        internal void WriteToStream(Stream destination)
-        {
-            destination.Write(WrittenMemory.Span);
-        }
+        internal void WriteToStream(Stream destination) => destination.Write(_buffer.ActiveSpan);
 #else
-        internal void WriteToStream(Stream destination)
-        {
-            Debug.Assert(_rentedBuffer != null);
-            destination.Write(_rentedBuffer, 0, _index);
-        }
+        internal void WriteToStream(Stream destination) => destination.Write(_buffer.DangerousGetUnderlyingBuffer(), _buffer.ActiveStartOffset, _buffer.ActiveLength);
 #endif
 
-        private void CheckAndResizeBuffer(int sizeHint)
-        {
-            Debug.Assert(_rentedBuffer != null);
-            Debug.Assert(sizeHint > 0);
-
-            int currentLength = _rentedBuffer.Length;
-            int availableSpace = currentLength - _index;
-
-            // If we've reached ~1GB written, grow to the maximum buffer
-            // length to avoid incessant minimal growths causing perf issues.
-            if (_index >= MaximumBufferSize / 2)
-            {
-                sizeHint = Math.Max(sizeHint, MaximumBufferSize - currentLength);
-            }
-
-            if (sizeHint > availableSpace)
-            {
-                int growBy = Math.Max(sizeHint, currentLength);
-
-                int newSize = currentLength + growBy;
-
-                if ((uint)newSize > MaximumBufferSize)
-                {
-                    newSize = currentLength + sizeHint;
-                    if ((uint)newSize > MaximumBufferSize)
-                    {
-                        ThrowHelper.ThrowOutOfMemoryException_BufferMaximumSizeExceeded((uint)newSize);
-                    }
-                }
-
-                byte[] oldBuffer = _rentedBuffer;
-
-                _rentedBuffer = ArrayPool<byte>.Shared.Rent(newSize);
-
-                Debug.Assert(oldBuffer.Length >= _index);
-                Debug.Assert(_rentedBuffer.Length >= _index);
-
-                Span<byte> oldBufferAsSpan = oldBuffer.AsSpan(0, _index);
-                oldBufferAsSpan.CopyTo(_rentedBuffer);
-                oldBufferAsSpan.Clear();
-                ArrayPool<byte>.Shared.Return(oldBuffer);
-            }
-
-            Debug.Assert(_rentedBuffer.Length - _index > 0);
-            Debug.Assert(_rentedBuffer.Length - _index >= sizeHint);
-        }
-
         public override async ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken = default)
         {
             Debug.Assert(_stream is not null);
 #if NET
             await _stream.WriteAsync(WrittenMemory, cancellationToken).ConfigureAwait(false);
 #else
-            Debug.Assert(_rentedBuffer != null);
-            await _stream.WriteAsync(_rentedBuffer, 0, _index, cancellationToken).ConfigureAwait(false);
+            await _stream.WriteAsync(_buffer.DangerousGetUnderlyingBuffer(), _buffer.ActiveStartOffset, _buffer.ActiveLength, cancellationToken).ConfigureAwait(false);
```

Comment:
I think we have an extension method somewhere that adds a ROM `WriteAsync` overload. Should we use this here as well?

Reply:
Thiking of this helper: https://github.com/dotnet/runtime/blob/d897415e02340a13dc1c5078c09937bdf7ec8a56/src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/Helpers.cs#L73-L97

Reply:
There were a couple more such helpers flying around, merged them to a common helper now.